frontend/app/performance/lighthouse/run.mjs

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+// @ts-check
+import fs from "fs"
+import path from "path"
+
+import { LighthouseOrchestrator } from "./LighthouseOrchestrator.mjs"
+import {
+  MODES,
+  PERFORMANCE_APPS_DIRECTORY,
+  STREAMLIT_ROOT,
+} from "./constants.mjs"
+
+/**
+ * Run the Lighthouse performance test suite.
+ * @param {string} appsDirectory The directory containing the streamlit apps to
+ * run
+ */
+const run = async appsDirectory => {
+  const now = new Date()
+    .toISOString()
+    .replace(/[-:.TZ]/g, "")
+    .slice(0, 15)
+    .replace("T", "-")
+  const runId = `${now.slice(0, 8)}-${now.slice(8, 14)}`
+
+  /**
+   * An array of performance app filenames.
+   * @type {string[]}
+   */
+  const performanceApps = fs
+    .readdirSync(appsDirectory, { withFileTypes: true })
+    .filter(entry => {
+      return (
+        !entry.isDirectory() &&
+        entry.name.endsWith(".py") &&
+        entry.name !== "__init__.py"
+      )
+    })
+    .map(({ name }) => name)
+
+  const orchestrator = new LighthouseOrchestrator()
+
+  process.on("SIGINT", async () => {
+    orchestrator.destroy()
+    // Exit with a non-zero status code to indicate that the process was
+    // interrupted.
+    process.exit(1)
+  })
+
+  await orchestrator.initialize()
+
+  try {
+    for (const appName of performanceApps) {
+      const appPath = path.join(appsDirectory, appName)
+      await orchestrator.startStreamlit(appPath, STREAMLIT_ROOT)
+
+      for (const mode of Object.keys(MODES)) {
+        await orchestrator.runLighthouse(appName, mode, runId)
+      }
+
+      await orchestrator.stopStreamlit()
+    }
+  } catch (e) {
+    console.error(e)
+    orchestrator.destroy()
+    // Exit with a non-zero status code to indicate that there was an error.
+    process.exit(1)
+  }
+
+  orchestrator.destroy()
+}
+
+// @ts-expect-error This is a top-level await that works fine in our version of Node
```

Comment:
Based on the fileending `.mjs` I wasn't aware that we use typescript here as well. If this is the case, perhaps it should be changed to `.mts` (if this is a thing 😄). Also, _if_ we have typescript compiling for these files, it might be nice to leverage type-hints for the function arguments and variables 🙂

Reply:
iirc part of typescript's adoption strategy was that you could add type support directly in javascript files via the `// @ts-check`. i feel using proper typescript here might be overkill needing to maintain the transpilation and such.

Reply:
Ah I did not know about the `// @ts-check`! I agree with the overhead of transpiling etc. 👍

Reply:
Yes, what Nico said is correct. I did not want to introduce transpiling, so `.mjs` + `// @ts-check` felt like the right balance here!